### DISCUSSION TITLE
Transform or middleware for modifying response bodies

### DISCUSSION THREAD

**Discussion by heathdavies-eaton** [dotnet/yarp#2389] [General] [answered]
I'm looking to modify a JSON response body before proxying it back to our client and first did this in a middleware. However after reading you documentation about what not to do in middleware it seems response modification is not advised in middleware.

https://microsoft.github.io/reverse-proxy/articles/middleware.html#what-not-to-do-with-middleware

I then read your section on response body transforms where response modification didn't seem to be discouraged as much

https://microsoft.github.io/reverse-proxy/articles/transforms.html#request-body-transforms

My question therefore is whether there it would be better to use middleware or a request body transform for this sort of modification?Or is there not really a difference?

**Comment by Tratcher** [comment, marked as answer]
Transforms are a little better because the content has to be copied from the HttpResponseMessage to the HttpResponse anyways, so inserting yourself in there isn't adding an extra step. If you can do the transformation in a streaming fashion (processing a bit at a time rather than reading the whole thing into memory) that's even better.

**Reply by socsieng** [reply]
I'm struggling to find sample code on how to implement this. Is there an existing example on the best way to implement this?
